File: src/jsyon/_global.py
from collections.abc import Sequence, Mapping
from .interpreter import Interpreter

NATIVE_SCOPE = {key: getattr(__builtins__, key) for key in dir(__builtins__)}

class Global:
    py = NATIVE_SCOPE
   
    @staticmethod
    def eval(body, context):
        return None

    zip = zip
    
    @staticmethod
    def eval(body, context):
        return None

    @staticmethod
    def print(*args):
        print(*args)
        return args

    Obj = lambda *args: dict(args)
    
    class Arr(list):
        def __init__(self, *args):
            super(Global.Arr, self).__init__(args)

        def push(self, item):
            self.append(item)
            return self

        def pop():
            return self.pop()

    @staticmethod
    def Fn(*args):
        body = args[-1];
        params = args[0:-1];
        return lambda *innerArgs: Interpreter(
            body,
            dict(zip(params, innerArgs))
        ).run()

    Num = float
    Str = str

File: src/native/debug/global.js
const Global = require('../global.js');
const Interpreter = require('./interpreter.js');
const Parser = require('../../parser.js');


class DebuggingGlobal extends Global {
    async eval(code, context) {
        return new Interpreter(new Parser(code).parse()).run(this, context);
    }

    async eval_json(json, context) {
        return new Interpreter(json).run(this, context);
    }

    Fn(...args) {
        let body = args[args.length - 1];
        let params = args.slice(0, args.length - 1);

        let fn = (...innerArgs) => {
            let context = this.Obj(...this.zip(params, innerArgs));
            context['__fn__'] = fn;
            context['__args__'] = innerArgs;
            context['__args_names__'] = params;
            return Interpreter.runExprGlobal(this, body, context);
        };
        return fn;
    }
}

module.exports = DebuggingGlobal;